### DISCUSSION TITLE
Stable JSON-serialization of sets to *ordered* lists

### DISCUSSION THREAD

**Discussion by pbasista** [pydantic/pydantic#8794] [Question]
Hi, I am wondering if there is a reasonably simple way to configure Pydantic 2 to always JSON-serialize (i.e serialize to *string* that can be parsed as JSON) sets to *ordered* lists.

In Pydantic 2, fields of type `set` are already JSON-serialized to lists (or JSON arrays). However, their items are ordered according to internal ordering of the original set. That produces unstable outcomes because even if two sets consist of the same items, their internal ordering might still be different.

In Pydantic 1, the desired effect could be achieved by using the [json_encoders](https://docs.pydantic.dev/1.10/usage/exporting_models/#json_encoders) parameter of the configuration and defining a custom serialization function for all attributes of type set. However, in Pydantic 2 this option has been [removed](https://docs.pydantic.dev/dev/migration/#changes-to-pydanticbasemodel) due to "performance overhead and implementation complexity". Which sounds understandable.

It seems that at some point of JSON serialization, Pydantic 2 is converting sets to lists anyway. Perhaps it is possible to somehow tap into this conversion and merely call `sorted` on its outcome.

**Comment by uriyyo** [comment]
@pbasista One of the possible solutions could be to introduce a custom set type, that will have a serializer that will going to sort values:
```py
from typing import Annotated, TypeVar

from pydantic import WrapSerializer

T = TypeVar("T")

SortedSet = Annotated[
    set[T],
    WrapSerializer(lambda v, next_: sorted(next_(v))),
]
```

Usage:
```py
from pydantic import BaseModel


class MyModel(BaseModel):
    s: SortedSet[int]


obj = MyModel(s={3, 2, 1})
print(obj.model_dump_json())
```
```py
{"s":[1,2,3]}
```

**Reply by pbasista** [reply]
Yes, I am aware of that possibility. But I think it is not the most user-friendly one. The reason is that I would then have to replace all `set` types with this new custom type.

But thank you anyway. It is definitely usable as one of the fallback options.
